### PR TITLE
[move-ide] Added support for variable shadowing

### PR DESCRIPTION
**Pull request by awelc**
## Description 

This PR adds handling of shadowed variables to the trace viewer:
- shadowed variables appear in separate scopes
- shadowed scope appears when a variable with the same name as an already existing one is declared
- shadowed scope disappears when all variables it includes stop being live

Additionally, this PR (temporarily) suppresses on-hover information that VSCode shows during debug session. The reason for it is that the default VSCode behavior in this department breaks in presence of shadowed variables

## Test plan 

Tested manually

### CODE REVIEW

File: external-crates/move/crates/move-analyzer/trace-adapter/src/runtime.ts [hunk 1]
```diff
@@ -433,13 +433,25 @@ export class Runtime extends EventEmitter {
...
                 for (let j = 0; j < currentFrame.locals[i].length; j++) {
                     if (frameLocalLifetimeEnds[j] === instructionEvent.pc) {
                         currentFrame.locals[i][j] = undefined;
                     }
                 }
             }
+            // trim shadowed scopes that have no live variables in them
+            if (localsLength > 1) {
+                for (let i = localsLength - 1; i > 0; i--) {
+                    const existingVar = currentFrame.locals[i].find(runtimeVar => {
```
Comment (dariorussi) [inline external-crates/move/crates/move-analyzer/trace-adapter/src/runtime.ts:447]: I'd probably call this `liveVar`? it matches the comment and the intent better?

File: external-crates/move/crates/move-analyzer/trace-adapter/src/runtime.ts [hunk 2]
```diff
@@ -433,13 +433,25 @@ export class Runtime extends EventEmitter {
...
+            const localsLength = currentFrame.locals.length;
+            for (let i = 0; i < localsLength; i++) {
                 for (let j = 0; j < currentFrame.locals[i].length; j++) {
                     if (frameLocalLifetimeEnds[j] === instructionEvent.pc) {
                         currentFrame.locals[i][j] = undefined;
                     }
                 }
             }
+            // trim shadowed scopes that have no live variables in them
+            if (localsLength > 1) {
```
Comment (dariorussi) [inline external-crates/move/crates/move-analyzer/trace-adapter/src/runtime.ts:445]: I am not sure I understand why we must have at least on local in the frame to enter the loop

**Review by dariorussi — COMMENTED** [review]
few questions as I will look at this some more

File: external-crates/move/crates/move-analyzer/trace-adapter/src/adapter.ts [hunk 3]
```diff
@@ -236,16 +236,15 @@ export class MoveDebugSession extends LoggingDebugSession {
...
-            scopes.push(localScope);
-            // TODO: finish shadowed variables support
-            for (let i = 1; i < frame.locals.length; i++) {
+            for (let i = frame.locals.length - 1; i > 0; i--) {
                 const shadowedScopeReference = this.variableHandles.create({ locals: frame.locals[i] });
                 const shadowedScope = new Scope(`shadowed(${i}): ${frame.name}`, shadowedScopeReference, false);
                 scopes.push(shadowedScope);
             }
         }
+        const localScopeReference = this.variableHandles.create({ locals: frame.locals[0] });
```
[the remark this answers is not in the document]
Comment (awelc) [reply external-crates/move/crates/move-analyzer/trace-adapter/src/adapter.ts:247]: There is indeed an implicit invariant at play here. We don't have to check if scope 0 exists as it's created whenever a new frame is create and it's never disposed of. I added a comment along those lines to the code

File: external-crates/move/crates/move-analyzer/trace-adapter/src/runtime.ts (hunk 2 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/trace-adapter/src/runtime.ts]: Indeed there is not need for this if condition. Removed!

File: external-crates/move/crates/move-analyzer/trace-adapter/src/runtime.ts (hunk 1 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/trace-adapter/src/runtime.ts]: Done!

**Review by tzakian — APPROVED** [review]
Looks good to me!

**Pull request merged**